src/core/coverage.ml

```diff
@@ -171,387 +120,469 @@ type coverage_result =
+(* Rigid matching algorithm : pre-matching algorithm which will generate
    candidates*)
 
-type depend   = Atomic | Dependent
+type depend =
+  | Atomic
+  | Dependent
 
 type cov_goal =
   | CovGoal of LF.dctx * LF.normal * LF.tclo (*  cPsi |- tR <= sP *)
   | CovCtx of LF.dctx
-  | CovSub of  LF.dctx * LF.sub * LF.cltyp
+  | CovSub of LF.dctx * LF.sub * LF.cltyp
   | CovPatt of gctx * Comp.pattern * Comp.tclo
 
 type pattern =
-  | MetaPatt  of  LF.dctx * LF.normal * LF.tclo
-  | MetaCtx   of LF.dctx
-  | MetaSub   of  LF.dctx * LF.sub * LF.cltyp
-  | EmptyPatt of  LF.dctx * LF.tclo
-  | EmptyParamPatt of  LF.dctx * LF.tclo
+  | MetaPatt of LF.dctx * LF.normal * LF.tclo
+  | MetaCtx of LF.dctx
+  | MetaSub of LF.dctx * LF.sub * LF.cltyp
+  | EmptyPatt of LF.dctx * LF.tclo
+  | EmptyParamPatt of LF.dctx * LF.tclo
   | EmptyCompPatt of Comp.tclo
   | GenPatt of Comp.gctx * Comp.pattern * Comp.tclo
 
+type eqn =
+  | Eqn of cov_goal * pattern
+  | EqnCtx of LF.dctx * LF.dctx
 
-type eqn   = Eqn of cov_goal * pattern | EqnCtx of LF.dctx * LF.dctx
-
-
-type split = Split of cov_goal * pattern | SplitCtx of LF.dctx * LF.dctx
-             | SplitPat of (Comp.pattern * Comp.tclo)  * (Comp.pattern * Comp.tclo)
+type split =
+  | Split of cov_goal * pattern
+  | SplitCtx of LF.dctx * LF.dctx
+  | SplitPat of (Comp.pattern * Comp.tclo) * (Comp.pattern * Comp.tclo)
 
 type candidate =
-    Cand of  LF.mctx *  Comp.gctx *        (* meta-context of pattern        *)
-              eqn list * split list
+  | Cand of
+      LF.mctx (* meta-context of pattern *)
+      * Comp.gctx
+      * eqn list
+      * split list
 
 type candidates = candidate list
 
-type covproblem =
+type cov_problem =
   LF.mctx * gctx * candidates * Comp.pattern
 
-type covproblems = covproblem list
+type cov_problems = cov_problem list
+
+type open_cov_problem =
+  LF.mctx * gctx * Comp.pattern
 
-let open_cov_goals  = ref ([]   :  (LF.mctx * gctx * Comp.pattern) list )
+type open_cov_problems = open_cov_problem list
 
-let reset_cov_problem () = (open_cov_goals := [])
+let open_cov_problems = ref ([] : open_cov_problems)
 
-type solved = Solved | NotSolvable | PossSolvable of candidate
+let reset_open_cov_problems () = open_cov_problems := []
+
+type solved =
+  | Solved
+  | NotSolvable
+  | PossSolvable of candidate
 
 type refinement_candidate =
   | TermCandidate of (LF.mctx * cov_goal * LF.msub)
   | CtxCandidate of (LF.mctx * LF.dctx * LF.msub)
-
 (*  | PatCandidate of (LF.mctx * gctx * cov_goal * LF.msub * Comp.pattern list) *)
 
 type refinement_cands =
-    NoCandidate
+  | NoCandidate
   | SomeTermCands of depend * (refinement_candidate list)
   | SomeCtxCands of refinement_candidate list
 (*  | SomePatCands of refinement_candidate list *)
 
 
-let rec lower cPsi sA = match sA with
-  | (LF.Atom (_ , a, _tS), s) -> (cPsi , Whnf.whnfTyp sA)
-  | (LF.PiTyp ((decl, _ ), tB), s) -> lower (LF.DDec (cPsi, S.LF.decSub decl s)) (tB, S.LF.dot1 s)
+let rec lower cPsi sA =
```

Comment:
Actually I think there is already an implementation of lowering somewhere else in the code. Perhaps we can call that instead of reimplementing it here.

Reply:
OK, I will try to find one.

Reply:
@tsani What does `lower` in this context mean?